NPE at aQute.bnd.osgi.Macro.process(Macro.java:96)

I'm getting NullPointerException by stacktrace below 
(Bnd is built from sources at revision 
SHA-1: 5780ccf68be8d0365f84ff385ed93632e6e17f5a
Merge pull request #1268 from bjhargrave/javac.defaults)

```
"main"
	at java.lang.StringBuilder.<init>(StringBuilder.java:125)
	at aQute.bnd.osgi.Macro.process(Macro.java:96)
	at aQute.bnd.osgi.Macro.process(Macro.java:91)
	at aQute.bnd.osgi.Macro.process(Macro.java:86)
	at aQute.bnd.osgi.Macro.process(Macro.java:1204)
	at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1152)
	at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1128)
	at aQute.bnd.osgi.Processor.getFlattenedProperties(Processor.java:1375)
	at aQute.bnd.osgi.Analyzer.calcManifest(Analyzer.java:881)
	at aQute.bnd.osgi.Builder.build(Builder.java:105)
	at aQute.bnd.osgi.Builder$build.call
```

It occurs because `process()` is invoked with `null` argument
```
	public Properties getFlattenedProperties(boolean ignoreInstructions) {
	      ...
						flattened.put(key, process(source.getProperty(key)));
```

what happens because of this code
```
package java.util;

public class Properties extends Hashtable<Object,Object> {
    public String getProperty(String key) {
        Object oval = super.get(key);
        String sval = (oval instanceof String) ? (String)oval : null;
        return ((sval == null) && (defaults != null)) ? defaults.getProperty(key) : sval;
    }
```

because value stored in Hashtable is actually a GString (a Groovy String) and is evaluated as null therefore (i would say Hashtable implementation is the origin of that problem). At the moment i have no idea how GString comes into Bnd's code (project is build by Gradle + plugin: 'org.dm.bundle''), but that stack trace with NPE is absolutely misleading for problem resolution and such weird condition should be handled more gracefully within Bnd code - for example name of the 'bad' key should be logged and more informative exception can be thrown.

The problem is actually not about GString only (groovy is dynamic language and does not perform sufficient type checks and it seems the problem is bit more severe than I expected originally - root exception occurs earlier in the code, here

```
package aQute.bnd.osgi;
public class Processor extends Domain implements Reporter, Registry, Constants, Closeable {
	public String getProperty(String key, String deflt, String separator) {
		...
				value = (String) source.getProperties().get(key);
```

what is invoked from
```
public class Analyzer extends Processor {
	public Manifest calcManifest() throws Exception {
		try {
		...
		}
		catch (Exception e) {
			// This should not really happen. The code should never throw
			// exceptions in normal situations. So if it happens we need more
			// information. So to help diagnostics. We do a full property dump
			throw new IllegalStateException("Calc manifest failed, state=\n" + getFlattenedProperties(), e);
		}
	}
```
And NPE exception occurs in the code which reports source exception :) ... and, BTW, after fixing source of GString i've got another one, with not GString:

```
Caused by: java.lang.IllegalArgumentException: Property Bnd-LastModified is not a not-null String, value: {java.lang.Long} 1450368311594
        at aQute.bnd.osgi.Macro.getFlattenedProperties(Macro.java:1155)
```

And here is the updated code (for copy/paste?), with more intelligent processing of that case 
```
public Properties getFlattenedProperties(boolean ignoreInstructions) {
        // Some macros only work in a lower processor, so we
        // do not report unknown macros while flattening
        flattening = true;
        try {
            Properties flattened = new UTF8Properties();
            Properties source = domain.getProperties();

            for( Enumeration< ?> e = source.propertyNames(); e.hasMoreElements(); ) {
                String key = (String)e.nextElement();
                if( !key.startsWith("_") ) {
                    String propValueString = source.getProperty(key);

                    if( propValueString == null ) {
                        Object propValue = source.get(key);
                        throw new IllegalArgumentException(String.format(
                                "Property %s is not a not-null String, value: %s%s",
                                key, propValue != null ? "{" + propValue.getClass().getName() + "} " : "", propValue));
                    }

                    if( ignoreInstructions && key.startsWith("-") )
                        flattened.put(key, propValueString);
                    else
                        flattened.put(key, process(propValueString));
                }
            }
            return flattened;
        }
        finally {
            flattening = false;
        }
    }
```

--- a/biz.aQute.bndlib.tests/src/test/MacroTest.java
+++ b/biz.aQute.bndlib.tests/src/test/MacroTest.java
@@ -1149,4 +1149,39 @@ public class MacroTest extends TestCase {
 		assertEquals("F31BAC7F1F70E5D8705B98CC0FBCFF5E", b.getProperty("b"));
 	}
 
+	public void testProcessNullValue() throws Exception {
+		try (Processor b = new Processor()) {
+			Macro m = b.getReplacer();
+			String tst = m.process(null);
+			assertEquals("", tst);
+			assertTrue(b.check());
+		}
+	}
+
+	public void testNonStringValue() throws Exception {
+		try (Processor b = new Processor()) {
+			// getProperty will return null for non-String value
+			b.getProperties().put("tst", new StringBuilder("foo"));
+			b.getProperties().put("num", 2);
+			String tst = b.getProperty("tst");
+			assertNull(tst);
+			String num = b.getProperty("num");
+			assertNull(num);
+			assertTrue(b.check("Key 'tst' has a non-String value", "Key 'num' has a non-String value"));
+		}
+	}
+
+	public void testNonStringFlattenedValue() throws Exception {
+		try (Processor b = new Processor()) {
+			// getProperty will return null for non-String value
+			b.getProperties().put("tst", new StringBuilder("foo"));
+			b.getProperties().put("num", 2);
+			Properties f = b.getFlattenedProperties();
+			String tst = f.getProperty("tst");
+			assertNull(tst);
+			String num = f.getProperty("num");
+			assertNull(num);
+			assertTrue(b.check("Key 'tst' has a non-String value", "Key 'num' has a non-String value"));
+		}
+	}
 }

--- a/biz.aQute.bndlib/src/aQute/bnd/osgi/Macro.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/osgi/Macro.java
@@ -93,6 +93,9 @@ public class Macro {
 	}
 
 	int process(CharSequence org, int index, char begin, char end, StringBuilder result, Link link) {
+		if (org == null) { // treat null like empty string
+			return index;
+		}
 		StringBuilder line = new StringBuilder(org);
 		int nesting = 1;
 
@@ -1145,11 +1148,19 @@ public class Macro {
 			Properties source = domain.getProperties();
 			for (Enumeration< ? > e = source.propertyNames(); e.hasMoreElements();) {
 				String key = (String) e.nextElement();
-				if (!key.startsWith("_"))
-					if (ignoreInstructions && key.startsWith("-"))
-						flattened.put(key, source.getProperty(key));
-					else
-						flattened.put(key, process(source.getProperty(key)));
+				if (!key.startsWith("_")) {
+					String value = source.getProperty(key);
+					if (value == null) {
+						Object raw = source.get(key);
+						domain.warning("Key '%s' has a non-String value: %s:%s", key,
+								raw == null ? "" : raw.getClass().getName(), raw);
+					} else {
+						if (ignoreInstructions && key.startsWith("-"))
+							flattened.put(key, value);
+						else
+							flattened.put(key, process(value));
+					}
+				}
 			}
 			return flattened;
 		}

--- a/biz.aQute.bndlib/src/aQute/bnd/osgi/Processor.java
+++ b/biz.aQute.bndlib/src/aQute/bnd/osgi/Processor.java
@@ -1204,12 +1204,27 @@ public class Processor extends Domain implements Reporter, Registry, Constants, 
 		// Use the key as is first, if found ok
 
 		if (filter != null && filter.contains(key)) {
-			value = (String) getProperties().get(key);
+			Object raw = getProperties().get(key);
+			if (raw != null) {
+				if (raw instanceof String) {
+					value = (String) raw;
+				} else {
+					warning("Key '%s' has a non-String value: %s:%s", key, raw == null ? "" : raw.getClass().getName(),
+							raw);
+				}
+			}
 		} else {
 			while (source != null) {
-				value = (String) source.getProperties().get(key);
-				if (value != null)
+				Object raw = source.getProperties().get(key);
+				if (raw != null) {
+					if (raw instanceof String) {
+						value = (String) raw;
+					} else {
+						warning("Key '%s' has a non-String value: %s:%s", key,
+								raw == null ? "" : raw.getClass().getName(), raw);
+					}
 					break;
+				}
 
 				source = source.getParent();
 			}

